File: apps/android/app/src/main/kotlin/jiamin/chen/orangecloud/ui/audit/AuditLogViewModel.kt
package jiamin.chen.orangecloud.ui.audit

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import jiamin.chen.orangecloud.core.auth.AuthRepository
import jiamin.chen.orangecloud.core.auth.Scopes
import jiamin.chen.orangecloud.data.model.AuditLogEntry
import jiamin.chen.orangecloud.data.repository.AccountStore
import jiamin.chen.orangecloud.data.repository.AuditLogRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.temporal.ChronoUnit
import javax.inject.Inject

data class AuditLogUiState(
    val entries: List<AuditLogEntry> = emptyList(),
    val isLoading: Boolean = false,
    val isLoadingMore: Boolean = false,
    val hasError: Boolean = false,
    val missingScope: Boolean = false,
    val hasMore: Boolean = false,
    /** 正在查看变更历史的那条日志（null 表示未打开）。 */
    val historyOf: AuditLogEntry? = null,
    val history: List<AuditLogEntry> = emptyList(),
    val isLoadingHistory: Boolean = false,
    /** exact / approximate / unavailable —— 识别质量，approximate 时必须提醒用户可能混入无关条目。 */
    val historyStatus: String? = null,
    val historyError: Boolean = false,
)

@HiltViewModel
class AuditLogViewModel @Inject constructor(
    private val accountStore: AccountStore,
    private val repository: AuditLogRepository,
    authRepository: AuthRepository,
) : ViewModel() {

    // 审计日志只需账号必选 scope（account-settings.read），通常恒有。
    private val hasScope = authRepository.hasScope(Scopes.ACCOUNT_READ)
    private var cursor: String? = null

    private val _uiState = MutableStateFlow(
        AuditLogUiState(isLoading = hasScope, missingScope = !hasScope),
    )
    val uiState: StateFlow<AuditLogUiState> = _uiState.asStateFlow()

    init {
        if (hasScope) loadFirst()
    }

    fun loadFirst() {
        if (!hasScope) return
        cursor = null
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, hasError = false, entries = emptyList()) }
            fetchPage(reset = true)
            _uiState.update { it.copy(isLoading = false) }
        }
    }

    fun loadMore() {
        if (!hasScope || cursor == null || _uiState.value.isLoadingMore) return
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingMore = true) }
            fetchPage(reset = false)
            _uiState.update { it.copy(isLoadingMore = false) }
        }
    }

    /** 打开某条日志对应资源的变更历史（用与列表相同的 30 天窗口）。 */
    fun openHistory(entry: AuditLogEntry) {
        val entryId = entry.id ?: return
        val actionTime = entry.timestampMillis?.let(Instant::ofEpochMilli) ?: return
        _uiState.update {
            it.copy(historyOf = entry, history = emptyList(), historyStatus = null,
                historyError = false, isLoadingHistory = true)
        }
        viewModelScope.launch {
            try {
                val accountId = accountStore.selectedAccountId.value ?: error("no account")
                val before = Instant.now()
                val since = before.minus(30, ChronoUnit.DAYS)
                val page = repository.resourceHistory(accountId, entryId, actionTime, since, before)
                _uiState.update {
                    it.copy(
                        history = page.result.orEmpty(),
                        historyStatus = page.resultInfo?.historyStatus,
                        isLoadingHistory = false,
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(historyError = true, isLoadingHistory = false) }
            }
        }
    }

    fun closeHistory() {
        _uiState.update { it.copy(historyOf = null, history = emptyList(), historyStatus = null) }
    }

    private suspend fun fetchPage(reset: Boolean) {
        try {
            accountStore.ensureLoaded()
            val accountId = accountStore.selectedAccountId.value ?: run {
                _uiState.update { it.copy(hasError = true) }
                return
            }
            val before = Instant.now()
            val since = before.minus(30, ChronoUnit.DAYS)
            val page = repository.list(accountId, since, before, cursor)
            cursor = page.cursor
            val entries = page.result.orEmpty()
            _uiState.update {
                it.copy(
                    entries = if (reset) entries else it.entries + entries,
                    hasMore = !page.cursor.isNullOrEmpty() && entries.isNotEmpty(),
                )
            }
        } catch (e: Exception) {
            _uiState.update { it.copy(hasError = true) }
        }
    }
}

File: apps/ios/Orange Cloud/Orange Cloud/ViewModels/AuditLogViewModel.swift
//
//  AuditLogViewModel.swift
//  Orange Cloud
//
//  审计日志列表：最近 30 天，游标分页（下滑续接）。
//

import Foundation
import Observation

@Observable
@MainActor
final class AuditLogViewModel {

    private(set) var entries: [IdentifiedAuditEntry] = []
    private(set) var canLoadMore = false
    var isLoading = false
    var isLoadingMore = false
    var error: String?

    private let service: AuditLogService
    private let accountId: String

    /// 时间窗在首次加载时固定，续页复用同一窗口（游标分页要求查询参数一致）
    private var since:  Date?
    private var before: Date?
    private var cursor: String?

    /// 查询窗口：最近 30 天
    private static let windowDays: TimeInterval = 30 * 24 * 3600

    init(service: AuditLogService, accountId: String) {
        self.service = service
        self.accountId = accountId
    }

    func load() async {
        guard !isLoading else { return }
        isLoading = true
        error = nil

        let before = Date()
        let since = before.addingTimeInterval(-Self.windowDays)
        self.before = before
        self.since = since
        self.cursor = nil

        do {
            let page = try await service.list(
                accountId: accountId, since: since, before: before, cursor: nil
            )
            entries = (page.result ?? []).map { IdentifiedAuditEntry(entry: $0) }
            cursor = page.cursor
            canLoadMore = !(page.cursor?.isEmpty ?? true)
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func loadMore() async {
        guard canLoadMore, !isLoading, !isLoadingMore,
              let cursor, let since, let before else { return }
        isLoadingMore = true
        do {
            let page = try await service.list(
                accountId: accountId, since: since, before: before, cursor: cursor
            )
            entries.append(contentsOf: (page.result ?? []).map { IdentifiedAuditEntry(entry: $0) })
            self.cursor = page.cursor
            canLoadMore = !(page.cursor?.isEmpty ?? true)
        } catch {
            self.error = error.localizedDescription
        }
        isLoadingMore = false
    }
    // MARK: - 资源变更历史

    private(set) var history: [IdentifiedAuditEntry] = []
    /// exact / approximate / unavailable，用于给用户交代结果可信度
    private(set) var historyStatus: String?
    var isLoadingHistory = false

    /// 拉某条日志所对应资源的变更序列。
    /// 时间窗沿用列表的 since/before —— 接口要靠 action_time 定位源日志，窗口不一致会找不到。
    func loadHistory(for entry: AuditLogEntry) async {
        guard let entryId = entry.id, let actionTime = entry.timestamp,
              let since, let before, !isLoadingHistory else { return }
        isLoadingHistory = true
        history = []
        historyStatus = nil
        do {
            let page = try await service.resourceHistory(
                accountId: accountId, entryId: entryId, actionTime: actionTime,
                since: since, before: before
            )
            history = (page.result ?? []).map { IdentifiedAuditEntry(entry: $0) }
            historyStatus = page.resultInfo?.historyStatus
        } catch {
            self.error = error.localizedDescription
        }
        isLoadingHistory = false
    }

}
